runtime/sema/errors.go

```diff
@@ -3026,6 +3027,71 @@ func (e *InvalidAccessError) Error() string {
 	)
 }
 
+// When e.PossessedAccess is a conjunctive entitlement set, we can suggest
+// which additional entitlements it would need to be given in order to have
+// e.RequiredAccess.
+func (e *InvalidAccessError) SecondaryError() string {
+	if e.PossessedAccess == nil || e.RestrictingAccess == nil {
+		return ""
+	}
+	possessedEntitlements, possessedOk := e.PossessedAccess.(EntitlementSetAccess)
+	requiredEntitlements, requiredOk := e.RestrictingAccess.(EntitlementSetAccess)
+	if !possessedOk && e.PossessedAccess.Equal(UnauthorizedAccess) {
+		possessedOk = true
+		// for this error reporting, model UnauthorizedAccess as an empty entitlement set
+		possessedEntitlements = NewEntitlementSetAccess([]*EntitlementType{}, Conjunction)
+	}
+	if !possessedOk || !requiredOk || possessedEntitlements.SetKind != Conjunction {
+		return ""
+	}
+
+	var sb strings.Builder
+
+	switch requiredEntitlements.SetKind {
+	case Conjunction:
+		// when both `possessed` and `required` are conjunctions, the missing set is simple set difference:
+		// `missing` = `required` - `possessed`, and `missing` should be added to `possessed` to make `required`
+		missingEntitlements := orderedmap.New[EntitlementOrderedSet](0)
+		requiredEntitlements.Entitlements.Foreach(func(key *EntitlementType, _ struct{}) {
+			if !possessedEntitlements.Entitlements.Contains(key) {
+				missingEntitlements.Set(key, struct{}{})
+			}
+		})
+		missingLen := missingEntitlements.Len()
+		if missingLen == 1 {
+			sb.WriteString("reference needs entitlement ")
+			sb.WriteString(fmt.Sprintf("`%s`", missingEntitlements.Newest().Key.QualifiedString()))
+		} else {
+			sb.WriteString("reference needs all of entitlements ")
+			missingEntitlements.ForeachWithIndex(func(index int, key *EntitlementType, _ struct{}) {
+				sb.WriteString(fmt.Sprintf("`%s`", key.QualifiedString()))
+				if index < missingLen-2 {
+					sb.WriteString(", ")
+				} else if index < missingLen-1 {
+					sb.WriteString(" and ")
```

Comment:
Here and below:

Sorry for being pedantic, maybe add Oxford comma if `missingEntitlements` has more than two elements.

There's a helper in `common.EnumerateWords`, but it allocates to a temporary string and requires a slice as input. Maybe it can be improved while at it and used here, to keep the logic here free of formatting details.